mdata/store_cassandra.go

```diff
@@ -73,13 +76,68 @@ It's safe for concurrent use by multiple goroutines and a typical usage scenario
 object to interact with the whole Cassandra cluster.
 */
 
+type ttlTables map[uint32]ttlTable
+type ttlTable struct {
+	table  string
+	window uint32
+}
+
 type cassandraStore struct {
 	session     *gocql.Session
 	writeQueues []chan *ChunkWriteRequest
 	readQueue   chan *ChunkReadRequest
+	ttlTables   ttlTables
+}
+
+func ttlUnits(ttl uint32) float64 {
+	// convert ttl to hours
+	return float64(ttl) / (60 * 60)
+}
+
+func getTTLTables(ttls []uint32, windowFactor int, nameFormat string) ttlTables {
+	tables := make(ttlTables)
+	for _, ttl := range ttls {
+		/*
+		 * the purpose of this loop is to bucket metrics of similar TTLs.
+		 * we first calculate the largest power of 2 that's smaller than the TTL and then divide the result by
+		 * the window factor. for example with a window factor of 20 we want to group the metrics like this:
+		 *
+		 * generated with: https://gist.github.com/replay/69ad7cfd523edfa552cd12851fa74c58
+		 *
+		 * +-----------------------------+---------------+-------------+----------+
+		 * |                hours / days |    table_name | window_size | sstables |
+		 * +-----------------------------+---------------+-------------+----------+
+		 * |                      1 / <1 |     metrics_1 |           1 |    1 - 1 |
+		 * |                  2 - 3 / <1 |     metrics_2 |           1 |    2 - 3 |
+		 * |                  4 - 7 / <1 |     metrics_4 |           1 |    4 - 7 |
+		 * |                 8 - 15 / <1 |     metrics_8 |           1 |   8 - 15 |
+		 * |                16 - 31 / ~1 |    metrics_16 |           1 |  16 - 31 |
+		 * |             32 - 63 / 1 - 3 |    metrics_32 |           2 |  16 - 31 |
+		 * |            64 - 127 / 2 - 6 |    metrics_64 |           4 |  16 - 31 |
+		 * |          128 - 255 / 5 - 11 |   metrics_128 |           7 |  18 - 36 |
+		 * |         256 - 511 / 10 - 22 |   metrics_256 |          13 |  19 - 39 |
+		 * |        512 - 1023 / 21 - 43 |   metrics_512 |          26 |  19 - 39 |
+		 * |       1024 - 2047 / 42 - 86 |  metrics_1024 |          52 |  19 - 39 |
+		 * |      2048 - 4095 / 85 - 171 |  metrics_2048 |         103 |  19 - 39 |
+		 * |     4096 - 8191 / 170 - 342 |  metrics_4096 |         205 |  19 - 39 |
+		 * |    8192 - 16383 / 341 - 683 |  metrics_8192 |         410 |  19 - 39 |
+		 * |  16384 - 32767 / 682 - 1366 | metrics_16384 |         820 |  19 - 39 |
+		 * | 32768 - 65535 / 1365 - 2731 | metrics_32768 |        1639 |  19 - 39 |
```

Comment:
Right. Actually, the value `19` came from the fact that the real number was floored, in reality it would be something like `19.56` which means the `ttl` spans effectively across `20` sstables. 

I changed the formula to handle `float`s properly: https://gist.github.com/replay/69ad7cfd523edfa552cd12851fa74c58#file-generate_window_size_comment_table-py-L31

Now it looks like that:

```
mst@mst-nb1:~$ python ./generate_window_size_comment_table.py 
+-------------------+---------------+-------------+----------+
| TTL hours (floor) |    table_name | window_size | sstables |
+-------------------+---------------+-------------+----------+
|                 1 |     metrics_1 |           1 |    1 - 1 |
|             2 - 3 |     metrics_2 |           1 |    2 - 3 |
|             4 - 7 |     metrics_4 |           1 |    4 - 7 |
|            8 - 15 |     metrics_8 |           1 |   8 - 15 |
|           16 - 31 |    metrics_16 |           1 |  16 - 31 |
|           32 - 63 |    metrics_32 |           2 |  16 - 32 |
|          64 - 127 |    metrics_64 |           4 |  16 - 32 |
|         128 - 255 |   metrics_128 |           7 |  19 - 37 |
|         256 - 511 |   metrics_256 |          13 |  20 - 40 |
|        512 - 1023 |   metrics_512 |          26 |  20 - 40 |
|       1024 - 2047 |  metrics_1024 |          52 |  20 - 40 |
|       2048 - 4095 |  metrics_2048 |         103 |  20 - 40 |
|       4096 - 8191 |  metrics_4096 |         205 |  20 - 40 |
|      8192 - 16383 |  metrics_8192 |         410 |  20 - 40 |
|     16384 - 32767 | metrics_16384 |         820 |  20 - 40 |
|     32768 - 65535 | metrics_32768 |        1639 |  20 - 40 |
+-------------------+---------------+-------------+----------+
```